flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java

```diff
@@ -137,6 +158,151 @@ public EqualTo equalTo(KeySelector<T2, KEY> keySelector)  {
+	 *
+	 * @param <IN1> Input type of elements from the first stream
+	 * @param <IN2> Input type of elements from the second stream
+	 * @param <KEY> The type of the key
+	 */
+	public static class TimeBounded<IN1, IN2, KEY> {
+
+		private static final String TIMEBOUNDED_JOIN_FUNC_NAME = "TimeBoundedJoin";
+
+		private final DataStream<IN1> left;
+		private final DataStream<IN2> right;
+
+		private final long lowerBound;
+		private final long upperBound;
+
+		private final KeySelector<IN1, KEY> keySelector1;
+		private final KeySelector<IN2, KEY> keySelector2;
+
+		private boolean lowerBoundInclusive;
+		private boolean upperBoundInclusive;
+
+		public TimeBounded(
+			DataStream<IN1> left,
+			DataStream<IN2> right,
+			long lowerBound,
+			long upperBound,
+			boolean lowerBoundInclusive,
+			boolean upperBoundInclusive,
+			KeySelector<IN1, KEY> keySelector1,
+			KeySelector<IN2, KEY> keySelector2) {
+
+			this.left = Preconditions.checkNotNull(left);
+			this.right = Preconditions.checkNotNull(right);
+
+			this.lowerBound = lowerBound;
+			this.upperBound = upperBound;
+
+			this.lowerBoundInclusive = lowerBoundInclusive;
+			this.upperBoundInclusive = upperBoundInclusive;
+
+			this.keySelector1 = Preconditions.checkNotNull(keySelector1);
+			this.keySelector2 = Preconditions.checkNotNull(keySelector2);
+		}
+
+		/**
+		 * Configure whether the upper bound should be considered exclusive or inclusive.
+		 */
+		public TimeBounded<IN1, IN2, KEY> upperBoundExclusive(boolean exclusive) {
+			this.upperBoundInclusive = !exclusive;
+			return this;
+		}
+
+		/**
+		 * Configure whether the lower bound should be considered exclusive or inclusive.
+		 */
+		public TimeBounded<IN1, IN2, KEY> lowerBoundExclusive(boolean exclusive) {
+			this.lowerBoundInclusive = !exclusive;
+			return this;
+		}
+
+		/**
+		 * Completes the join operation with the user function that is executed for each joined pair
+		 * of elements.
+		 * @param udf The user-defined function
+		 * @param <OUT> The output type
+		 * @return Returns a DataStream
+		 */
+		public <OUT> DataStream<OUT> process(TimeBoundedJoinFunction<IN1, IN2, OUT> udf) {
+
+			ConnectedStreams<IN1, IN2> connected = left.connect(right);
+
+			udf = left.getExecutionEnvironment().clean(udf);
+
+			TypeInformation<OUT> resultType = TypeExtractor.getBinaryOperatorReturnType(
+				udf,
+				TimeBoundedJoinFunction.class,    // TimeBoundedJoinFunction<IN1, IN2, OUT>
+				0,                                //						  0    1    2
+				1,
+				2,
+				new int[]{0},                   // lambda input 1 type arg indices
+				new int[]{1},                   // lambda input 1 type arg indices
+				TypeExtractor.NO_INDEX,         // output arg indices
+				left.getType(),                 // input 1 type information
+				right.getType(),                // input 1 type information
+				TIMEBOUNDED_JOIN_FUNC_NAME,
+				false
+			);
+
+			long bucketGranularity = Time.seconds(1).toMilliseconds();
```

Comment:
does the granularity need to be documented?

Reply:
Yes, this should actually be exposed either through the API or config file (or both?). I'll give this one some thought as well and come back

Reply:
I would be up for not exposing this parameter to the users. This is an internal thing that we use in the current implementation. In the future, we may not even use buckets for the joins.

In addition, exposing it means exposing internal parameters to the user, which normally is not a good idea. We can always add this parameter if we see that it makes a difference in practice. Removing it if we change our mind in the future is always more tricky.